### PR TITLE
google-map-react - Update interface for Maps

### PR DESCRIPTION
**Pull request by tetarchus**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [Google Maps API V3 Reference](https://developers.google.com/maps/documentation/javascript/reference)
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## google-map-react ([<kbd>unpkg</kbd>](https://unpkg.com/browse/google-map-react@latest/))
was missing the following properties:
1. propTypes
1. defaultProps
1. googleMapLoader
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against d9eea37166033dc0d7ea4caeff730359b1d1b6ec
</p>

File: types/google-map-react/index.d.ts [hunk 1]
```diff
@@ -66,30 +66,191 @@ declare namespace googleMapReact {
...
+            RIGHT_TOP: number;
+            RIGHT: number;
+            RIGHT_CENTER: number;
+            RIGHT_BOTTOM: number;
+            BOTTOM_LEFT: number;
+            BOTTOM_CENTER: number;
+            BOTTOM: number;
+            BOTTOM_RIGHT: number;
+            CENTER: number;
+        };
```
Comment (romain-faust) [inline types/google-map-react/index.d.ts:93]: Could not find `TOP`, `LEFT`, `RIGHT`, `BOTTOM` and `CENTER` in [the documentation](https://developers.google.com/maps/documentation/javascript/reference/control#ControlPosition). Where do they come from ?

File: types/google-map-react/index.d.ts [hunk 2]
```diff
@@ -66,30 +66,191 @@ declare namespace googleMapReact {
...
-        TransitMode: any;
-        TransitRoutePreference: any;
-        TravelMode: any;
-        UnitSystem: any;
+        Animation: {
+            BOUNCE: number;
+            DROP: number;
+            Uz: number;
+            Qz: number;
+        };
```
Comment (romain-faust) [inline types/google-map-react/index.d.ts:74]: Could not find `Uz` and `Qz` in [the documentation](https://developers.google.com/maps/documentation/javascript/reference/marker#Animation). Where do they come from ?

File: types/google-map-react/index.d.ts [hunk 3]
```diff
@@ -66,30 +66,191 @@ declare namespace googleMapReact {
...
+            LESS_WALKING: string;
+            FEWER_TRANSFERS: string;
+        };
+        TravelMode: {
+            DRIVING: string;
+            WALKING: string;
+            BICYCLING: string;
+            TRANSIT: string;
+            TWO_WHEELER: string;
+        };
```
Comment (romain-faust) [inline types/google-map-react/index.d.ts:243]: Could not find `TWO_WHEELER` in [the documentation](https://developers.google.com/maps/documentation/javascript/reference/directions#TravelMode). Where does it come from ?

File: types/google-map-react/index.d.ts [hunk 4]
```diff
@@ -66,30 +66,191 @@ declare namespace googleMapReact {
...
+            MAX_WAYPOINTS_EXCEEDED: string;
+            NOT_FOUND: string;
+        };
+        DirectionsTravelMode: {
+            DRIVING: string;
+            WALKING: string;
+            BICYCLING: string;
+            TRANSIT: string;
+            TWO_WHEELER: string;
+        };
```
Comment (romain-faust) [inline types/google-map-react/index.d.ts:110]: Could not find `DirectionsTravelMode` in [the documentation](https://developers.google.com/maps/documentation/javascript/reference) (it seems to be the same as `TravelMode` though). Maybe it should be deprecated with a message telling users to use `TravelMode`, wdyt ?

File: types/google-map-react/index.d.ts [hunk 5]
```diff
@@ -66,30 +66,191 @@ declare namespace googleMapReact {
...
+        DirectionsUnitSystem: {
+            METRIC: number;
+            IMPERIAL: number;
+        };
```
Comment (romain-faust) [inline types/google-map-react/index.d.ts:114]: Could not find `DirectionsUnitSystem` in [the documentation](https://developers.google.com/maps/documentation/javascript/reference) (it seems to be the same as `UnitSystem` though). Maybe it should be deprecated with a message telling users to use `UnitSystem`, wdyt ?

File: types/google-map-react/index.d.ts [hunk 6]
```diff
@@ -66,30 +66,191 @@ declare namespace googleMapReact {
...
+            ZERO_RESULTS: string;
+        };
+        ElevationStatus: {
+            OK: string;
+            UNKNOWN_ERROR: string;
+            OVER_QUERY_LIMIT: string;
+            REQUEST_DENIED: string;
+            INVALID_REQUEST: string;
+            yz: string;
+        };
```
Comment (romain-faust) [inline types/google-map-react/index.d.ts:136]: Could not find `yz` in [the documentation](https://developers.google.com/maps/documentation/javascript/reference/elevation#ElevationStatus). Where does it come from ?

File: types/google-map-react/index.d.ts [hunk 7]
```diff
@@ -66,30 +66,191 @@ declare namespace googleMapReact {
...
+            LIMITS_EXCEEDED: string;
+            TIMED_OUT: string;
+        };
+        MapTypeControlStyle: {
+            DEFAULT: number;
+            HORIZONTAL_BAR: number;
+            DROPDOWN_MENU: number;
+            INSET: number;
+            INSET_LARGE: number;
+        };
```
Comment (romain-faust) [inline types/google-map-react/index.d.ts:169]: Could not find `INSET` and `INSET_LARGE` in [the documentation](https://developers.google.com/maps/documentation/javascript/reference/control#MapTypeControlStyle). Where do they come from ?

File: types/google-map-react/index.d.ts [hunk 8]
```diff
@@ -66,30 +66,191 @@ declare namespace googleMapReact {
...
+            ERROR: string;
+        };
+        NavigationControlStyle: {
+            DEFAULT: number;
+            SMALL: number;
+            ANDROID: number;
+            ZOOM_PAN: number;
+            Vz: number;
+            Ot: number;
+        };
```
Comment (romain-faust) [inline types/google-map-react/index.d.ts:187]: Could not find `NavigationControlStyle` in [the documentation](https://developers.google.com/maps/documentation/javascript/reference). Where does it come from ?

File: types/google-map-react/index.d.ts [hunk 9]
```diff
@@ -66,30 +66,191 @@ declare namespace googleMapReact {
...
+        UnitSystem: {
+            METRIC: number;
+            IMPERIAL: number;
+        };
+        ZoomControlStyle: {
+            DEFAULT: number;
+            SMALL: number;
+            LARGE: number;
+            Ot: number;
+        };
```
Comment (romain-faust) [inline types/google-map-react/index.d.ts:253]: Could not find `ZoomControlStyle` in [the documentation](https://developers.google.com/maps/documentation/javascript/reference). Where does it come from ?

**Review by romain-faust — CHANGES_REQUESTED** [review]
Thank you for taking the time to submit this PR @tetarchus 🙏 . I made small comments, don't hesitate to tell me if you've any questions. Do you think that it should be possible to just order values alphabetically (in the Google documentation keys are sorted alphabetically and it would facilitate the review 🙂 ) ?

**Comment by tetarchus** [comment]
@romain-faust Thanks for reviewing. The values of these were taken directly from a `console.log()` of the `Maps` object in a `createMapOptions` function suggested in the `google-map-react` documentation [here](https://github.com/google-map-react/google-map-react/blob/master/API.md#options-funcobject). It appeared that the options are essentially enums, so could be typed exactly as below, but I've been a little more generic to allow for changes to the values.
(https://github.com/google-map-react/google-map-react/blob/master/API.md#ontilesloaded-func)
```js
<GoogleMap  onGoogleApiLoaded={({map, maps}) => console.log(map, maps)} />
```

To replicate, just add a `console.log(maps)` as below:

```js
createMapOptions: function (maps) {
    console.log(maps);

    return {
      panControl: false,
      mapTypeControl: false,
      scrollwheel: false,
      styles: [{ stylers: [{ 'saturation': -100 }, { 'gamma': 0.8 }, { 'lightness': 4 }, { 'visibility': 'on' }] }]
    }
  }

 <GoogleMap  options={createMapOptions} ... />
```

This then returns the following object:

```JSON
{
    "Animation": {
        "BOUNCE": 1,
        "DROP": 2,
        "Uz": 3,
        "Qz": 4
    },
    "ControlPosition": {
        "TOP_LEFT": 1,
        "TOP_CENTER": 2,
        "TOP": 2,
        "TOP_RIGHT": 3,
        "LEFT_CENTER": 4,
        "LEFT_TOP": 5,
        "LEFT": 5,
        "LEFT_BOTTOM": 6,
        "RIGHT_TOP": 7,
        "RIGHT": 7,
        "RIGHT_CENTER": 8,
        "RIGHT_BOTTOM": 9,
        "BOTTOM_LEFT": 10,
        "BOTTOM_CENTER": 11,
        "BOTTOM": 11,
        "BOTTOM_RIGHT": 12,
        "CENTER": 13
    },
    "DirectionsStatus": {
        "OK": "OK",
        "UNKNOWN_ERROR": "UNKNOWN_ERROR",
        "OVER_QUERY_LIMIT": "OVER_QUERY_LIMIT",
        "REQUEST_DENIED": "REQUEST_DENIED",
        "INVALID_REQUEST": "INVALID_REQUEST",
        "ZERO_RESULTS": "ZERO_RESULTS",
        "MAX_WAYPOINTS_EXCEEDED": "MAX_WAYPOINTS_EXCEEDED",
        "NOT_FOUND": "NOT_FOUND"
    },
    "DirectionsTravelMode": {
        "DRIVING": "DRIVING",
        "WALKING": "WALKING",
        "BICYCLING": "BICYCLING",
        "TRANSIT": "TRANSIT",
        "TWO_WHEELER": "TWO_WHEELER"
    },
    "DirectionsUnitSystem": {
        "METRIC": 0,
        "IMPERIAL": 1
    },
    "DistanceMatrixStatus": {
        "OK": "OK",
        "INVALID_REQUEST": "INVALID_REQUEST",
        "OVER_QUERY_LIMIT": "OVER_QUERY_LIMIT",
        "REQUEST_DENIED": "REQUEST_DENIED",
        "UNKNOWN_ERROR": "UNKNOWN_ERROR",
        "MAX_ELEMENTS_EXCEEDED": "MAX_ELEMENTS_EXCEEDED",
        "MAX_DIMENSIONS_EXCEEDED": "MAX_DIMENSIONS_EXCEEDED"
    },
    "DistanceMatrixElementStatus": {
        "OK": "OK",
        "NOT_FOUND": "NOT_FOUND",
        "ZERO_RESULTS": "ZERO_RESULTS"
    },
    "ElevationStatus": {
        "OK": "OK",
        "UNKNOWN_ERROR": "UNKNOWN_ERROR",
        "OVER_QUERY_LIMIT": "OVER_QUERY_LIMIT",
        "REQUEST_DENIED": "REQUEST_DENIED",
        "INVALID_REQUEST": "INVALID_REQUEST",
        "yz": "DATA_NOT_AVAILABLE"
    },
    "GeocoderLocationType": {
        "ROOFTOP": "ROOFTOP",
        "RANGE_INTERPOLATED": "RANGE_INTERPOLATED",
        "GEOMETRIC_CENTER": "GEOMETRIC_CENTER",
        "APPROXIMATE": "APPROXIMATE"
    },
    "GeocoderStatus": {
        "OK": "OK",
        "UNKNOWN_ERROR": "UNKNOWN_ERROR",
        "OVER_QUERY_LIMIT": "OVER_QUERY_LIMIT",
        "REQUEST_DENIED": "REQUEST_DENIED",
        "INVALID_REQUEST": "INVALID_REQUEST",
        "ZERO_RESULTS": "ZERO_RESULTS",
        "ERROR": "ERROR"
    },
    "KmlLayerStatus": {
        "UNKNOWN": "UNKNOWN",
        "OK": "OK",
        "INVALID_REQUEST": "INVALID_REQUEST",
        "DOCUMENT_NOT_FOUND": "DOCUMENT_NOT_FOUND",
        "FETCH_ERROR": "FETCH_ERROR",
        "INVALID_DOCUMENT": "INVALID_DOCUMENT",
        "DOCUMENT_TOO_LARGE": "DOCUMENT_TOO_LARGE",
        "LIMITS_EXCEEDED": "LIMITS_EXECEEDED",
        "TIMED_OUT": "TIMED_OUT"
    },
    "MapTypeControlStyle": {
        "DEFAULT": 0,
        "HORIZONTAL_BAR": 1,
        "DROPDOWN_MENU": 2,
        "INSET": 3,
        "INSET_LARGE": 4
    },
    "MapTypeId": {
        "ROADMAP": "roadmap",
        "SATELLITE": "satellite",
        "HYBRID": "hybrid",
        "TERRAIN": "terrain"
    },
    "MaxZoomStatus": {
        "OK": "OK",
        "ERROR": "ERROR"
    },
    "NavigationControlStyle": {
        "DEFAULT": 0,
        "SMALL": 1,
        "ANDROID": 2,
        "ZOOM_PAN": 3,
        "Vz": 4,
        "Ot": 5
    },
    "RenderingType": {
        "UNINITIALIZED": "UNINITIALIZED",
        "RASTER": "RASTER",
        "VECTOR": "VECTOR"
    },
    "ScaleControlStyle": {
        "DEFAULT": 0
    },
    "StreetViewPreference": {
        "NEAREST": "nearest",
        "BEST": "best"
    },
    "StreetViewStatus": {
        "OK": "OK",
        "UNKNOWN_ERROR": "UNKNOWN_ERROR",
        "ZERO_RESULTS": "ZERO_RESULTS"
    },
    "StreetViewSource": {
        "DEFAULT": "default",
        "OUTDOOR": "outdoor"
    },
    "StrokePosition": {
        "CENTER": 0,
        "INSIDE": 1,
        "OUTSIDE": 2
    },
    "SymbolPath": {
        "CIRCLE": 0,
        "FORWARD_CLOSED_ARROW": 1,
        "FORWARD_OPEN_ARROW": 2,
        "BACKWARD_CLOSED_ARROW": 3,
        "BACKWARD_OPEN_ARROW": 4
    },
    "TrafficModel": {
        "BEST_GUESS": "bestguess",
        "OPTIMISTIC": "optimistic",
        "PESSIMISTIC": "pessimistic"
    },
    "TransitMode": {
        "BUS": "BUS",
        "RAIL": "RAIL",
        "SUBWAY": "SUBWAY",
        "TRAIN": "TRAIN",
        "TRAM": "TRAM"
    },
    "TransitRoutePreference": {
        "LESS_WALKING": "LESS_WALKING",
        "FEWER_TRANSFERS": "FEWER_TRANSFERS"
    },
    "TravelMode": {
        "DRIVING": "DRIVING",
        "WALKING": "WALKING",
        "BICYCLING": "BICYCLING",
        "TRANSIT": "TRANSIT",
        "TWO_WHEELER": "TWO_WHEELER"
    },
    "UnitSystem": {
        "METRIC": 0,
        "IMPERIAL": 1
    },
    "ZoomControlStyle": {
        "DEFAULT": 0,
        "SMALL": 1,
        "LARGE": 2,
        "Ot": 3
    },
    "event": {}
}
```

Which was used to create the types - the order was based on the order it returned them at (which kinda makes sense for the numerical enums, not so much for the string enums). I have re-ordered these alphabetically however.

File: types/google-map-react/index.d.ts (hunk 1 above)
Comment (tetarchus) [reply types/google-map-react/index.d.ts]: Please see [main comment](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/60329#issuecomment-1126798899) for response.

File: types/google-map-react/index.d.ts (hunk 2 above)
Comment (tetarchus) [reply types/google-map-react/index.d.ts]: Please see [main comment](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/60329#issuecomment-1126798899) for response.

File: types/google-map-react/index.d.ts (hunk 3 above)
Comment (tetarchus) [reply types/google-map-react/index.d.ts]: Please see [main comment](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/60329#issuecomment-1126798899) for response.

File: types/google-map-react/index.d.ts (hunk 4 above)
Comment (tetarchus) [reply types/google-map-react/index.d.ts]: Please see [main comment](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/60329#issuecomment-1126798899) for response RE where this was located. 

In terms of being deprecated, it seems that these were renamed back in 2011 but still persist for backwards compatability (see [Release Notes](https://developers.google.com/maps/documentation/javascript/releases?hl=en#april-14,-2011)) - quoted below:

> Renamed DirectionsTravelMode and DirectionsUnitSystem to TravelMode and UnitSystem (old names remain backwards compatible) 

As such it likely makes sense to mark as deprecated with the message - I'll make that change.

File: types/google-map-react/index.d.ts (hunk 5 above)
Comment (tetarchus) [reply types/google-map-react/index.d.ts]: Please see [previous comment](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/60329#discussion_r873068127)

File: types/google-map-react/index.d.ts (hunk 6 above)
Comment (tetarchus) [reply types/google-map-react/index.d.ts]: Please see [main comment](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/60329#issuecomment-1126798899) for response.

File: types/google-map-react/index.d.ts (hunk 7 above)
Comment (tetarchus) [reply types/google-map-react/index.d.ts]: Please see [main comment](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/60329#issuecomment-1126798899) for response.

File: types/google-map-react/index.d.ts (hunk 8 above)
Comment (tetarchus) [reply types/google-map-react/index.d.ts]: Please see [main comment](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/60329#issuecomment-1126798899) for response on where it comes from. 

The only mention I can find of this is [here](https://developers.google.com/maps/documentation/javascript/releases?hl=en#feb-26,-2010) where it mentions that the ZOOM_PAN option was not working in IE8 back in 2010, but there are no other mentions of this option in the documentation at all (even that it's deprecated, or been removed from the docs!)

File: types/google-map-react/index.d.ts (hunk 9 above)
Comment (tetarchus) [reply types/google-map-react/index.d.ts]: Please see [main comment](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/60329#issuecomment-1126798899) for response on where it comes from. 

This appears to be depreacated and I will mark it as such [source](https://developers.google.com/maps/documentation/javascript/releases?hl=en#15-september-2015)

**Comment by romain-faust** [comment]
@tetarchus thank you for your answer, I think that we should keep the undocumented values because someone might use them.

File: types/google-map-react/index.d.ts [hunk 10]
```diff
@@ -66,30 +66,197 @@ declare namespace googleMapReact {
...
+        TransitRoutePreference: {
+            FEWER_TRANSFERS: string;
+            LESS_WALKING: string;
+        };
+        TravelMode: {
+            BICYCLING: string;
+            DRIVING: string;
+            WALKING: string;
+            TRANSIT: string;
+            TWO_WHEELER: string;
```
Comment (romain-faust) [inline types/google-map-react/index.d.ts:244]:
```suggestion
            TRANSIT: string;
            TWO_WHEELER: string;
            WALKING: string;
```

Think that you forgot those properties when sorting alphabetically 😉

**Review by romain-faust — CHANGES_REQUESTED** [review]
1 small comment otherwise LGTM 👍

File: types/google-map-react/index.d.ts (hunk 10 above)
Comment (tetarchus) [reply types/google-map-react/index.d.ts]: 🤦🏼

**Review by romain-faust — APPROVED** [review]
LGTM 👍

**Comment by tetarchus** [comment]
Ready to merge

**Pull request merged**